File: Problems/Graphs/Longest-Consecutive-Sequence/Longest-Consecutive-Sequence.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;

// Using union find algorithm - Time: O(n)

public class UnionFind
{
    private int[] id;
    private int[] size;

    public UnionFind(int n)
    {
        id = new int[n];
        size = new int[n];

        for (int i = 0; i < n; i++)
        {
            id[i] = i;
            size[i] = 1;
        }
    }

    public void Connect(int a, int b)
    {
        int x = Find(a);
        int y = Find(b);
        if (x == y) return;
        id[x] = y;
        size[y] += size[x];
    }

    public int Find(int a)
    {
        return id[a] == a ? a : (id[a] = Find(id[a]));
    }

    public int[] GetSizes()
    {
        return size;
    }
}

public class Solution
{
    public int LongestConsecutive(int[] nums)
    {
        if (nums.Length == 0) return 0;

        UnionFind uf = new UnionFind(nums.Length);
        Dictionary<int, int> m = new Dictionary<int, int>();

        for (int i = 0; i < nums.Length; i++)
        {
            int n = nums[i];
            if (m.ContainsKey(n)) continue;
            m[n] = i;
            if (m.ContainsKey(n + 1)) uf.Connect(m[n], m[n + 1]);
            if (m.ContainsKey(n - 1)) uf.Connect(m[n], m[n - 1]);
        }

        return uf.GetSizes().Max();
    }
}

class Program
{
    static void Main()
    {
        int[] nums = { 0,3,7,2,5,8,4,6,0,1 };
        Console.WriteLine("Input: [" + string.Join(",", nums) + "]");

        Solution sol = new Solution();
        int result = sol.LongestConsecutive(nums);

        Console.WriteLine("Output: " + result);
    }
}

File: Problems/Graphs/Longest-Consecutive-Sequence/Longest-Consecutive-Sequence.c
#include <stdio.h>
#include <stdlib.h>

// Using union find algorithm - Time: O(n)

struct UnionFind
{
    int *id;
    int *size;
};

void initializeUnionFind(struct UnionFind *uf, int n)
{
    uf->id = (int *)malloc(n * sizeof(int));
    uf->size = (int *)malloc(n * sizeof(int));

    for (int i = 0; i < n; i++)
    {
        uf->id[i] = i;
        uf->size[i] = 1;
    }
}

int find(struct UnionFind *uf, int a)
{
    return (uf->id[a] == a) ? a : (uf->id[a] = find(uf, uf->id[a]));
}

void connect(struct UnionFind *uf, int a, int b)
{
    int x = find(uf, a);
    int y = find(uf, b);

    if (x == y)
        return;

    uf->id[x] = y;
    uf->size[y] += uf->size[x];
}

int *getSizes(struct UnionFind *uf)
{
    return uf->size;
}

void freeUnionFind(struct UnionFind *uf)
{
    free(uf->id);
    free(uf->size);
}

int longestConsecutive(int *nums, int numsSize)
{
    if (numsSize == 0)
        return 0;

    struct UnionFind uf;
    initializeUnionFind(&uf, numsSize);

    int *m = (int *)malloc(numsSize * sizeof(int));

    for (int i = 0; i < numsSize; i++)
    {
        int n = nums[i];
        if (m[n])
            continue;

        m[n] = i;

        if (m[n + 1])
            connect(&uf, m[n], m[n + 1]);

        if (m[n - 1])
            connect(&uf, m[n], m[n - 1]);
    }

    int *sizes = getSizes(&uf);
    int result = 0;

    for (int i = 0; i < numsSize; i++)
    {
        if (sizes[i] > result)
            result = sizes[i];
    }

    free(m);
    freeUnionFind(&uf);

    return result;
}

int main()
{
    int nums[] = {0,3,7,2,5,8,4,6,0,1};
    int numsSize = sizeof(nums) / sizeof(nums[0]);

    printf("Input: [");
    for (int i = 0; i < numsSize; i++)
    {
        printf("%d", nums[i]);
        if (i < numsSize - 1)
            printf(",");
    }
    printf("]\n");

    int result = longestConsecutive(nums, numsSize);

    printf("Output: %d\n", result);

    return 0;
}
